File: lt0333.cpp
#include <algorithm>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x):val(x),left(nullptr), right(nullptr){}
};


class Solution{
public:
    int largetBSTSubtree(TreeNode* root){
        bool res = true;
        int mval = INT_MIN;
        return largestBSTSubtree(root, res, mval);
    }


    int largestBSTSubtree(TreeNode* root, bool& res, int& mval){
        if(root == nullptr){
            res = true;
            return 0;
        }

        mval = max(root->val, mval);

        if(root->left == nullptr && root->right == nullptr){
            res = true;
            return 1;
        }

        int left_nodes = 0, right_nodes = 0, lmvl = INT_MIN, rmvl = INT_MIN;

        left_nodes = largestBSTSubtree(root->left, res, lmvl);

        if(res)
            right_nodes = largestBSTSubtree(root->right, res, rmvl);


        if(res && root->left && (root->val < root->left->val || root->val < lmvl))
            res = false;
        
        if(res && root->right && (root->val > root->right->val || root->val > rmvl))
            res = false;

        if(res)
            left_nodes = left_nodes + right_nodes + 1;

        return max(left_nodes, right_nodes);        
    }
};

File: lt0333.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

 
class Solution {
    public int largestBSTSubtree(TreeNode root) {
        helper(root);
        return maxNode;    
    }
    
    public int[] helper(TreeNode node){
        if(node == null){
            return new int[]{0, Integer.MAX_VALUE, Integer.MIN_VALUE};
        }
        
        int[] left = helper(node.left);
        int[] right = helper(node.right);
        
        int maxV = left[2];
        int minV = right[1];
        
        if(left[0] == -1 || right[0] == -1)
            return new int[]{-1, -1, -1};
            
        boolean isBST = true;
        
        if(node.left!=null){
            isBST = node.val > maxV;
        }
            
        if(isBST && node.right != null){
            isBST = node.val < minV;
        }
        
        maxV = Math.max(right[2], node.val);
        minV = Math.min(left[1], node.val);
                
        int res = isBST ? left[0] + right[0] + 1 : -1;
        maxNode = Math.max(res, maxNode);
        return new int[]{res, minV, maxV};
        
    }
     
    private int maxNode = 0;
}